pulsar-broker/src/main/java/org/apache/pulsar/broker/service/ConsistentHashingStickyKeyConsumerSelector.java

```diff
@@ -53,55 +56,127 @@ public ConsistentHashingStickyKeyConsumerSelector(int numberOfPoints) {
+            if (index == -1) {
+                index = bitmap.getCardinality();
+            }
+            bitmap.add(index);
+            return index;
+        }
+
+        private void deallocateConsumerNameIndex(String consumerName, int index) {
+            RoaringBitmap bitmap = getConsumerNameIndexBitmap(consumerName);
+            bitmap.remove(index);
+            if (bitmap.isEmpty()) {
+                consumerNameCounters.remove(consumerName);
+            }
+        }
+
+        public void removeHashRingReference(ConsumerIdentityWrapper removed) {
+            ConsumerEntry consumerEntry = consumerEntries.get(removed);
+            int refCount = consumerEntry.refCount.decrementAndGet();
+            if (refCount == 0) {
+                deallocateConsumerNameIndex(consumerEntry.consumerName, consumerEntry.nameIndex);
+                consumerEntries.remove(removed, consumerEntry);
+            }
+        }
+
+        public int addHashRingReference(ConsumerIdentityWrapper wrapper) {
+            String consumerName = wrapper.consumer.consumerName();
+            ConsumerEntry entry = consumerEntries.computeIfAbsent(wrapper,
+                    k -> new ConsumerEntry(consumerName, allocateConsumerNameIndex(consumerName),
+                            new MutableInt(0)));
+            entry.refCount.increment();
+            return entry.nameIndex;
+        }
+
+        public int getTrackedConsumerNameIndex(ConsumerIdentityWrapper wrapper) {
+            ConsumerEntry consumerEntry = consumerEntries.get(wrapper);
+            return consumerEntry != null ? consumerEntry.nameIndex : -1;
+        }
+    }
+
     @Override
     public CompletableFuture<Void> addConsumer(Consumer consumer) {
         rwLock.writeLock().lock();
         try {
+            ConsumerIdentityWrapper consumerIdentityWrapper = new ConsumerIdentityWrapper(consumer);
             // Insert multiple points on the hash ring for every consumer
             // The points are deterministically added based on the hash of the consumer name
             for (int i = 0; i < numberOfPoints; i++) {
-                int hash = calculateHashForConsumerAndIndex(consumer, i);
-                hashRing.compute(hash, (k, v) -> {
-                    if (v == null) {
-                        return Lists.newArrayList(consumer);
-                    } else {
-                        if (!v.contains(consumer)) {
-                            v.add(consumer);
-                            v.sort(Comparator.comparing(Consumer::consumerName, String::compareTo));
-                        }
-                        return v;
-                    }
-                });
+                int consumerNameIndex = consumerNameIndexTracker.addHashRingReference(consumerIdentityWrapper);
+                int hash = calculateHashForConsumerAndIndex(consumer, consumerNameIndex, i);
+                ConsumerIdentityWrapper removed = hashRing.put(hash, consumerIdentityWrapper);
```

Comment:
> In my understanding is that the current approach "attempts to create different keys even if the consumerName is the same. However, in this approach, when hashes collide, it is replaced by the one added later." Is this understanding correct?

@equanz Yes, that's correct. However the hash collision won't happen due to using the same consumer name since there will be a unique key for each consumer with the same consumer name. 
When the previous is replaced, it won't cause existing consumers to switch for a specific hash. 

Any sort of implementation where there's a sub list or data structure for colliding consumers will result in uneven distribution since each hash range has a different size. That's why all entries must be in a single hash ring for the Key_shared Pulsar use case. I have made extensive experimentation and research to back this. If anyone is interested in more details, I can share them. I shared some details my previous comment, https://github.com/apache/pulsar/pull/23327#issuecomment-2376288786.

Reply:
> the hash collision won't happen due to using the same consumer name since there will be a unique key consumerNameIndex for each consumer with the same consumer name.

I see. Thank you.

I think this PR no longer guarantees against coincidental hash collisions referred to in the https://github.com/apache/pulsar/pull/8396.
However, it depends on the definition of what should be guaranteed(e.g. always consistent, almost consistent and always distributed).
What do you think?

Reply:
> I think this PR no longer guarantees against coincidental hash collisions referred to in the #8396.
> However, it depends on the definition of what should be guaranteed(e.g. always consistent, almost consistent and always distributed).
> What do you think?

The previous problem with collisions with the consumer name was that all consumers had collisions on all hash ring points (100 by default). That is solved in this PR by ensuring that when consumers have the same consumer name, there's an additional key, `consumerNameIndex`. That's how there won't be collisions on all hash ring points. Individual collisions aren't a problem since there will be 100 hash ring points by default. If there's a single collision, the other hash ring points will ensure that there won't be any problems in practice. I don't think that there's any drawbacks on consistency because of the 100 hash ring points. If one consumer has 1 point less due to a collision, the impact is statistically in the range of a few percentages. 

Even without any changes coming from this PR, the distribution is already not accurately distributed with 100 hash ring points. Currently the error margin is roughly in the range of 20-30% with 100 hash ring points. With 200 points, the accuracy improves and variation reduces, (IIRC about 15% in my experimentation). Also from this view point, having hash collisions on single hash ring points doesn't impact the accuracy significantly.

Reply:
Thnaks for your description.
I probably already know your point, but it is not my concern. My concern is the definition of ConsistentHashing's guarantee of consistency. If we change the definition of consistency, at least we should add some comments about it.

By the way, in my opinion, the above(changing the definition of consistency) is not a problem in Key_Shared since Key_Shared does not guarantee that the hash range is consistent regardless of the connection order. In fact, the HashRangeAutoSplitStickyKeyConsumerSelector depends on the connection order.
And the above is not a problem in ConsistentHashingStickyKeyConsumerSelector. In https://github.com/apache/pulsar/pull/6791 the consistent hashing is introduced for load balancing reasons. It is not intended to be consistent regardless of connection order.

I think this PR is a great work.

Reply:
> The previous problem with collisions with the consumer name was that all consumers had collisions on all hash ring points (100 by default). That is solved in this PR by ensuring that when consumers have the same consumer name, there's an additional key, consumerNameIndex. That's how there won't be collisions on all hash ring points. Individual collisions aren't a problem since there will be 100 hash ring points by default. If there's a single collision, the other hash ring points will ensure that there won't be any problems in practice. I don't think that there's any drawbacks on consistency because of the 100 hash ring points. If one consumer has 1 point less due to a collision, the impact is statistically in the range of a few percentages.

@lhotari But user can also set the points to 1 for some reason since it's configurable. Or even a smaller value, not one. If we can accept this case, the hash collisions from different consumer names should also be fine, just replace the old point, like the old way without https://github.com/apache/pulsar/pull/8396

My point is that we can make the change more simple

1. Change the consumer key, have consumer_name, consumer_id, address (it still work with non-proxy case)
2. Just have a rehash method for the conflict point

Something like this:
```java
for (int i = 0; i < numberOfPoints; i++) {
    int hash = calculateHashForConsumerAndIndex(consumer, i);
    Consumer existConsumer = hashRing.putIfAbsent(hash, consumer);
    while (existConsumer != null) {
        hash = rehash(hash);
        existConsumer = hashRing.putIfAbsent(hash, consumer);
    }
}
```

Both replace or rehash has it's pros and cons, and the cons for both of them are not a big problem.

- Replace, we are leveraging the points and message keys for the load distribution. One replace might also cause the existing consumer cannot get any messages.
- Rehash, consumer will not always added to the same point when hash collisions happened. But it should be fine, it's only a tiny part of all the consumers.

So, I prefer a less change, straightforward solution.

@lhotari

Reply:
But I don't want to say it's a blocker for this PR. @lhotari If you still want to continue with your solution in this PR, I'm fine with it. I can continue to review the details.

Reply:
> My point is that we can make the change more simple
> 
> 1. Change the consumer key, have consumer_name, consumer_id, address (it still work with non-proxy case)

Adding consumer_id will break the current intended behavior that when a consumer disconnects and reconnects immediately, that it gets assigned with the same hash ranges as before the disconnection happened. When using unique consumer names, the behavior change would be a significant regression. There's no way to determine beforehand whether a consumer name will be unique or not. That's why the logic would need to be used in all cases.

Adding address isn't useful at all for most users of Pulsar. The socket ports aren't useful since a new socket would be used in most cases after reconnecting after a connection is terminated and re-established. It wouldn't be an effective solution.

> 2. Just have a rehash method for the conflict point

Makes sense for handling the conflict.

> So, I prefer a less change, straightforward solution.

Please point to exact details of what is "much change" if you say that there's a lot of changes in this PR. It will help understand your point. Perhaps there's a way to improve this PR to match your criteria. Let's stay open for that. This PR is addressing a bug fix with a change that minimizes the risk of regressions. I believe that it's a common goal for us.

Reply:
> My point is that we can make the change more simple
> Change the consumer key, have consumer_name, consumer_id, address (it still work with non-proxy case)
Just have a rehash method for the conflict point
> Something like this: ...

Agree with @codelipenghui

Reply:
> Agree with @codelipenghui, we can choose a simpler implementation if we just want to fix the out-of-order issue.

@poorbarcode this has been addressed in the most recent changes. If you disagree about the simplicity, please provide specific feedback which part of the code in this PR is too complex in your mind.

Reply:
> And the major part is the rehash solution is more straightforward. Readers can read it and understand it in 1 min without understand why it's safe for replacing the point,

@codelipenghui I have refactored the PR to follow this rule. Please take another look. 
My expectation is that you could point to exact details in the code of this PR if there are concerns that the code is too complex. I hope that is a reasonable ask from my side.